### PR TITLE
[SPARK-3726] [MLlib] Allow sampling_rate not equal to 1.0 in RandomForests

### PR DESCRIPTION
**Pull request by MechCoder**
I've added support for sampling_rate not equal to 1.0 . I have two major questions.
1. A Scala style test is failing, since the number of parameters now exceed 10.
2. I would like suggestions to understand how to test this.

### CODE REVIEW

**Comment by MechCoder** [comment]
@jkbradley @mengxr it would be great if you could have a look.

**Comment by MechCoder** [comment]
I've made changes such that this not break anything, i.e everything is backward compat.

**Comment by jkbradley** [comment]
You can run dev/scalastyle locally to see what the issues are.

**Comment by MechCoder** [comment]
@jkbradley, the issue is that the function `train` has more than 10 args.

**Comment by MechCoder** [comment]
@jkbradley Oops, the comments got deleted somehow. I meant that this is because there are 10 arguments in `trainClassifier` and `trainRegressor`

**Comment by jkbradley** [comment]
@MechCoder  Taking a closer look, I now realize that part of this functionality is already there...see the JIRA & let me know what you think.

**Comment by MechCoder** [comment]
Oh well, but still if I'm not mistaken, the `subSamplingRate` is overriden by the condition `numTrees > 1`.  This should not be the case as having a lower sampling, might help in this case too (when using the entire data might be expensive). IMO, That should be fixed, right?

**Comment by jkbradley** [comment]
Good point, yes, I think it's worth fixing.

**Comment by jkbradley** [comment]
Also, as far as testing....it's hard.  One way might be to:
* Run RF with a random seed and subsampling rate 1.0
* Run it the same way, but with with rate < 1.0
* Make sure it learns different forests.
* To make this test robust, you'll need to use the right (small but not too small) dataset size (maybe 5 features and 20 instances?), and also use a fixed random seed.

**Comment by MechCoder** [comment]
Thanks, Also a design decision, is it worthy enough to add this as an option to `train` given that it is now within the "style limit"?

**Comment by jkbradley** [comment]
I'd vote for not adding it to train since that part of the API is so unwieldy.

**Comment by MechCoder** [comment]
Could you please tell me what is the preferred way to generate random data in spark?

**Comment by MechCoder** [comment]
@jkbradley I've added a test according to the other tests in the `RandomForestSuite` . Let me know if there is anything left.

**Comment by MechCoder** [comment]
@mengxr @jkbradley Any more comments? Sorry for spamming, but I would like to work on other issues related to GBRT and RandomForests as well.

File: mllib/src/test/scala/org/apache/spark/mllib/tree/RandomForestSuite.scala [hunk 1]
```diff
@@ -196,6 +196,24 @@ class RandomForestSuite extends FunSuite with MLlibTestSparkContext {
...
+      useNodeIdCache = true, subsamplingRate=0.5)
+    val strategy2 = new Strategy(algo = Classification, impurity = Gini, maxDepth = 2,
+      numClasses = 2, categoricalFeaturesInfo = Map.empty[Int, Int],
+      useNodeIdCache = true)
+
+    val rf1 = RandomForest.trainClassifier(rdd, strategy1, numTrees = 3,
+      featureSubsetStrategy = "auto", seed = 123)
+    val rf2 = RandomForest.trainClassifier(rdd, strategy2, numTrees = 3,
+      featureSubsetStrategy = "auto", seed = 123)
+    assert(rf1.totalNumNodes != rf2.totalNumNodes)
```
Comment (MechCoder) [inline mllib/src/test/scala/org/apache/spark/mllib/tree/RandomForestSuite.scala]: @jkbradley I've thought about this. This does not seem an ideal test. i.e If this condition satisfies, then the forests are not the same, but there might be cases where the number of nodes are still the same but the forests are not equal. Do you have any better ideas?

Comment (jkbradley) [inline mllib/src/test/scala/org/apache/spark/mllib/tree/RandomForestSuite.scala]: @MechCoder I'll try to make another pass now; it's been busy!

For the test, I'd compare trees by printing toDebugString and comparing the strings.  That should test for exact equality.

File: mllib/src/main/scala/org/apache/spark/mllib/tree/RandomForest.scala [hunk 2]
```diff
@@ -132,6 +132,7 @@ private class RandomForest (
     timer.start("init")
 
     val retaggedInput = input.retag(classOf[LabeledPoint])
+    val subsample = strategy.subsamplingRate
```
Comment (jkbradley) [inline mllib/src/main/scala/org/apache/spark/mllib/tree/RandomForest.scala]: I would just eliminate this temp value.

File: mllib/src/test/scala/org/apache/spark/mllib/tree/RandomForestSuite.scala [hunk 3]
```diff
@@ -196,6 +196,24 @@ class RandomForestSuite extends FunSuite with MLlibTestSparkContext {
       featureSubsetStrategy = "sqrt", seed = 12345)
     EnsembleTestHelper.validateClassifier(model, arr, 1.0)
   }
+
+  test("subsampling rate in RandomForest"){
+    val arr = EnsembleTestHelper.generateOrderedLabeledPoints(5, 20)
+    val rdd = sc.parallelize(arr)
+    val strategy1 = new Strategy(algo = Classification, impurity = Gini, maxDepth = 2,
```
Comment (jkbradley) [inline mllib/src/test/scala/org/apache/spark/mllib/tree/RandomForestSuite.scala]: I would make 1 instance of strategy, train rf1, modify the strategy's subsamplingRate, and train rf2.  Simpler + more clearly using the same settings for other parameters

**Comment by MechCoder** [comment]
@jkbradley Thanks for the tip. Fixed. Anything more?

**Comment by MechCoder** [comment]
Repushed after fixing the style checks.

**Comment by MechCoder** [comment]
ping @jkbradley Could you please have a final look?

**Comment by jkbradley** [comment]
@MechCoder   This is an addition instead of a correction, but I just realized that Strategy.assertValid() does not check subsamplingRate.  Would you mind adding that check?  The rest looks good to me.  Thanks!

**Comment by MechCoder** [comment]
@jkbradley Fixed. I can haz merge?

**Comment by jkbradley** [comment]
@MechCoder Thanks!  LGTM

CC: @mengxr  Note this is sort of an API change: RandomForest can now be run with subsampled rows.  (But this seems fine to me since users could set subsamplingRate before---it just wouldn't do anything.)

**Comment by MechCoder** [comment]
@mengxr This can also be viewd as a bugfix which prevents overwriting of the param `subSamplingRate`, which was hardcoded to 1.0

**Comment by mengxr** [comment]
Merged into master. Thanks!

**Pull request closed without merging**